### ISSUE TITLE
No module named 'mliap_model_python_couple_kokkos'

### ISSUE THREAD

**Issue by hoba87** [FitSNAP/FitSNAP#188]
Not sure if this is a fitsnap or lammps problem, however when I try to run a fitsnap pytorch model I get this error. Using a linear fitsnap model is working fine.
I'm using the recent lammps version from 08 Feb 2023. The command for lammps I used is
`mpiexec -n 1 lmp -k on g 1 -pk kokkos newton on neigh half -sf kk -in in.lammps`
Without active kokkos I get strange pointer errors.

Full error message:
ERROR: Loading MLIAPPYKokkos coupling module failure. (src/KOKKOS/mliap_model_python_kokkos.cpp:66)
Last command: `pair_style mliap model mliappy FitTorch_Pytorch.pt descriptor sna TiO2_snap.mliap.descriptor`
No module named 'mliap_model_python_couple_kokkos'

**Comment by rohskopf** [comment]
This might be more of a LAMMPS problem but let's keep this issue up.

Kokkos MLIAP is undergoing some changes.

@hoba87 Are you able to run the native LAMMPS MLIAP examples with Kokkos? Particularly the `lammps/examples/mliap` directory in the LAMMPS repo.

**Comment by athomps** [comment]
Yes, the MLIAP+Kokkos+PyTorch setup in LAMMPS currently only works for running MD, it does not provide support for FitSNAP training. The fact that you got it to work for linear is probably because it "fell-back" to non-Kokkos.

**Comment by athomps** [comment]
So, you should have FitSNAP invoke non-Kokkos LAMMPS. @rohskopf can help with the pointer errors.

**Comment by rohskopf** [comment]
There is also a Python example script specifically for Kokkos: https://github.com/lammps/lammps/blob/develop/examples/mliap/mliap_pytorch_Ta06A_kokkos.py

**Comment by hoba87** [comment]
Ok, I got the examples to run, with some small code fixes, look here: https://github.com/lammps/lammps/pull/3664
However I just were able to run the python script, due to the loading of the mliap_model_python_couple_kokkos module.
I also can run it with 4 mpi processes (kokkos option gpu/aware off) which increases the computation speed slightly (however e.g. 8 mpi processes gives me errors).
@athomps I have trained a model and now I am trying to use it for MD.
With my own example, I can also run it with 4 mpi processes, but they are evenly distributed among 1 CPU, not 4 (perhaps some GIL issue?) and nearly no GPU usage (model layer sizes were defined by `num_desc 60 60 1`) giving me a 59KB model file. Should that be the case? Therefore I have no increased computation speed, but also no decrease with 4 cpus compared to 1.

Besides that I get many of this ignored exceptions (tested with pytorch 1.8.2 and 1.12.1):
```
Exception ignored in: 'mliap_model_python_couple_kokkos.MLIAPPYKokkos_compute_gradients'
Traceback (most recent call last):
  File "/opt/python/python-3.7.9/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
TypeError: forward() takes 5 positional arguments but 6 were given
```
and probably only KOKKOS related after the calculation:
`Kokkos::Cuda ERROR: Failed to call Kokkos::Cuda::finalize()`

**Comment by athomps** [comment]
Yes, the Kokkos/MLIAP combination is very new, and is mainly targeting force evaluation on GPUs for PyTorch models defined by `unified` rather than `model`.  This effort is being led by @bathmatt and there is an open PR https://github.com/lammps/lammps/pull/3661

**Comment by bathmatt** [comment]
I'm interested in resolving this, there is very limited testing available in lammps for this and I'm really not surprised there are problems.  The patch that @athomps mentions I don't believe will address this as it addresses issues in the unified model.  I can look into this a bit tomorrow and then work through the chain to get this into the mainline.  It might take up to a week though.

**Comment by rohskopf** [comment]
Thanks @bathmatt. Maybe we can resolve the limited testing/examples with models that we generate here. Lemme know if you'd like some example potentials.

**Comment by bathmatt** [comment]
If I can test your problem w/ and w/o kokkos and see the difference that
would be helpful.  My guess is the interface changed since i copied over
the starting point.  I can look a bit at change logs and try to fix but it
might be helpful to have an example.  I'm on travel Th-We so tomorrow I can
spend a bit of time but then it will be a while

**Comment by bathmatt** [comment]
OK, I understand the issue.  I'm not sure how to deal with this issue and
am open to suggestions.

With the couple in the kokkos branch there was special code to handle
device pointers, this is the argument that was added.  I understand that
this caused issues with other plugins and it is out of my limited testing
(I never heard of this project honestly but this isn't my community).

How to move forward, my suggestion is to bring the non-kokkos in line to
match the kokkos version with a hardcoded false.  Other option is that we
can add a switch to lammps to use the smaller number of args like in the
legacy interface and only have the new one if we are passing cuda
pointers.

Pointer introspection is another option.  cudaPointerGetAttributes is the
C++ call, can this be done in python?  If that is the case we can remove
the dev pointer and clean up things a lot.

The last option is my preferred if I can figure out how to do so.

**Comment by rohskopf** [comment]
Looks like that is supported in CUDA Python:
https://nvidia.github.io/cuda-python/module/cuda.html#cuda.cuda.cuPointerGetAttribute

Should be straightforward to include that when deploying our models to LAMMPS. If it's convenient to keep the `dev` pointer though, we can just include that when deploying models on our end.

**Comment by bathmatt** [comment]
I am thinking not passing it in but computing that in the function.

**Comment by hoba87** [comment]
@bathmatt If you need my example for testing, I have uploaded it here:
[test_lammps_result_2.zip](https://github.com/FitSNAP/FitSNAP/files/10942141/test_lammps_result_2.zip)

**Comment by bathmatt** [comment]
I am looking at having a meeting with LANL to have a standard interface
between GPU and CPU and I'd like to pull you in as well.  And Aidan most
likely.  More info to come
